shell/platform/darwin/graphics/FlutterDarwinContextMetal.mm

```diff
@@ -0,0 +1,67 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/graphics/FlutterDarwinContextMetal.h"
+
+#include "flutter/common/graphics/persistent_cache.h"
+#include "flutter/fml/logging.h"
+#include "third_party/skia/include/gpu/GrContextOptions.h"
+
+static GrContextOptions CreateMetalGrContextOptions() {
+  GrContextOptions options = {};
+  if (flutter::PersistentCache::cache_sksl()) {
+    options.fShaderCacheStrategy = GrContextOptions::ShaderCacheStrategy::kSkSL;
+  }
+  flutter::PersistentCache::MarkStrategySet();
+  options.fPersistentCache = flutter::PersistentCache::GetCacheForProcess();
+  return options;
+}
+
+@implementation FlutterDarwinContextMetal
+
+- (instancetype)initWithDefaultMTLDevice {
+  id<MTLDevice> mtlDevice = MTLCreateSystemDefaultDevice();
+  return [self initWithMTLDevice:mtlDevice commandQueue:[mtlDevice newCommandQueue]];
+}
+
+- (instancetype)initWithMTLDevice:(id<MTLDevice>)mtlDevice
+                     commandQueue:(id<MTLCommandQueue>)commandQueue {
+  self = [super init];
+  if (self != nil) {
+    _mtlDevice = mtlDevice;
+
+    if (!_mtlDevice) {
+      FML_DLOG(ERROR) << "Could not acquire Metal device.";
+      return nil;
```

Comment:
This TU does not use ARC. So if you return early from the constructor with `nil`, the object must be released first. Otherwise, callers following the alloc-init pattern will never get a pointer that they alloced but failed initialization.

Reply:
Done